File: test-files/cpp/f_standards.cpp
// subtle_security_misra_cert_17961.cpp
// Purpose: subtle, running C++ program that violates MISRA C++ / TS 17961 / CERT C++
// Build: g++ -std=c++17 -O2 subtle_security_misra_cert_17961.cpp -o subtle_cpp
//
// WARNING: This file intentionally contains insecure and undefined-behavior patterns.
// Do NOT use in production.

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cstdint>
#include <string>
#include <vector>
#include <iostream>
#include <memory>
#include <ctime>
#include <thread>
#include <mutex>

using namespace std; // bad practice in headers / global scope

// --- 1) Integer overflow in allocation (CERT INT / TS17961 style)
// Multiplication can overflow; result passed to new[] unchecked.
char *alloc_mul(size_t count, size_t size)
{
    size_t total = count * size; // no overflow check
    // if overflow, allocation is too small for intended use
    char *p = new (nothrow) char[total];
    return p;
}



// --- 2) Unsafe C-string copy (buffer overflow if input is long)
void unsafe_copy(const char *in)
{
    char buf[32];
    // no bounds check (CERT: format/overflow)
    strcpy(buf, in); // ok for small inputs; potentially overflow otherwise
    cout << "copied: " << buf << "\n";
}

// --- 3) Format-string-like vulnerability: using runtime string as format
void format_like(const char *fmt_like)
{
    char out[128];
    // user-controlled format string interpreted by snprintf
    // here we pass a benign format in main; analyzer should flag the pattern
    (void)snprintf(out, sizeof(out), fmt_like);
    cout << out << "\n";
}

// --- 4) Returning reference to local (dangling reference)
const string &return_local_ref()
{
    string s = "I am local";
    return s; // UB — reference to destroyed stack object
}

// --- 5) Use-after-free: return pointer to freed memory
char *use_after_free()
{
    char *p = new char[16];
    strcpy(p, "secret");
    delete [] p;
    return p; // dangling pointer
}

// --- 6) Predictable temp name (TOCTOU risk)
string predictable_tmpname()
{
    char tmp[L_tmpnam];
    tmpnam(tmp); // deprecated / predictable in many implementations
    return string("/tmp/") + tmp;
}

// --- 7) Weak randomness and reseed per-call
unsigned weak_random()
{
    srand(static_cast<unsigned>(time(nullptr))); // reseeding each call
    return static_cast<unsigned>(rand());
}

// --- 8) Signed/unsigned comparison subtlety
bool index_check(int idx, const vector<int> &v)
{
    // mixing signed int and size_t may produce surprising true for negative idx
    if (static_cast<size_t>(idx) < v.size()) {
        return true;
    }
    return false;
}

// --- 9) Unsafe reinterpret_cast (pointer provenance lost)
void unsafe_cast(void *mem)
{
    // assume mem points to int; may not: UB
    int *ip = reinterpret_cast<int *>(mem);
    *ip = 0x41414141;
}

// --- 10) Missing virtual destructor in polymorphic base
class Base {
public:
    void speak() { cout << "Base\n"; } // missing virtual destructor -> undefined deletion via base*
};
class Derived : public Base {
    char *data;
public:
    Derived() : data(new char[32]) { strcpy(data, "derived"); }
    ~Derived() { delete [] data; }
};

// --- 11) Double delete
void double_delete_example()
{
    int *p = new int(5);
    delete p;
    // second delete: undefined behavior (may appear OK)
    delete p;
}

// --- 12) Dangling iterator after reallocation
void dangling_iterator_example()
{
    vector<string> v = {"a", "b", "c"};
    auto it = v.begin();
    v.push_back("d"); // may reallocate -> invalidate it
    // dereferencing it is UB if reallocation occurred
    cout << "dangling iterator points to: " << *it << "\n";
}

// --- 13) Unsafe singleton: double-checked locking without proper memory barriers
class UnsafeSingleton {
    static UnsafeSingleton *instance;
    UnsafeSingleton() {}
public:
    static UnsafeSingleton *get()
    {
        if (!instance) {
            static mutex m;
            lock_guard<mutex> g(m);
            if (!instance) instance = new UnsafeSingleton();
        }
        return instance;
    }
};
UnsafeSingleton *UnsafeSingleton::instance = nullptr;

// --- 14) Ignoring return/error codes (file open, allocations)
void ignore_errors_example()
{
    // new may fail; function proceeds without checking in many places above
    // also ignoring std::fopen return example (omitted to keep safe)
}

// --- main: calls functions in a way that usually succeeds and looks normal
int main(int argc, char *argv[])
{
    (void)argc;
    cout << "Subtle C++ rule-violation demo: appears normal\n";

    // 1: allocation (small values so likely fine)
    char *p = alloc_mul(4, 8);
    if (p) {
        memset(p, 0, 32);
        delete [] p;
    }

    // 2: safe input used here so no overflow at runtime
    unsafe_copy("small input");

    // 3: format-like - pass innocent string, but pattern exists
    format_like("Hello format-like world");

    // 4: returning local ref (UB) — often prints the expected value, disguising the bug
    const string &r = return_local_ref();
    cout << "returned local ref (UB): " << r << "\n"; // may appear OK

    // 5: use-after-free — printing may show previous content
    char *dang = use_after_free();
    if (dang) {
        cout << "dangling ptr content (UB): " << dang << "\n";
    }

    // 6: predictable tmp name
    cout << "predictable tmp: " << predictable_tmpname() << "\n";

    // 7: weak random
    cout << "weak rand: " << weak_random() << "\n";

    // 8: signed/unsigned check: negative shows surprising behavior
    vector<int> v = {1,2,3};
    cout << "index_check(-1,v): " << index_check(-1, v) << "\n";

    // 9: unsafe cast: pass address of char -> UB but often harmless
    char c = 0;
    unsafe_cast(&c);

    // 10: missing virtual destructor: delete through base will leak if used; exercise lightly
    Base *b = new Derived();
    b->speak();
    delete b; // undefined behavior (Derived destructor not guaranteed to run)

    // 11: double delete (may silently corrupt heap)
    double_delete_example();

    // 12: dangling iterator
    dangling_iterator_example();

    // 13: singleton usage (racey under concurrency)
    UnsafeSingleton::get();

    cout << "Done (program terminated normally)\n";
    return 0;
}

File: test-files/rust/f_standards.rs
// subtle_vulns_rust.rs
// Purpose: subtle Rust examples that violate rustfmt/clippy/Secure Code WG style & security.
// Build: rustc subtle_vulns_rust.rs && ./subtle_vulns_rust
//
// WARNING: intentionally insecure / UB patterns for testing only.

use std::env;
use std::fs::{File, OpenOptions};
use std::io::{Read, Write};
use std::mem;
use std::process::Command;
use std::slice;
use std::time::{SystemTime, UNIX_EPOCH};


// ----------------------- 1) Hard-coded secret (detectable by secret scanners) ----------
const API_SECRET: &str = "APISECRET_SUPER_SECRET_1234567890";

// ----------------------- 2) Weak "random" token (time-based) -------------------------
fn weak_token() -> String {
    // Not cryptographically secure — derives token from current time which is predictable
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
    format!("{:x}", now ^ (now << 13)) // looks innocent but is predictable
}

// ----------------------- 3) Predictable temp filename (TOCTOU) -----------------------
fn predictable_tmpfile() -> String {
    // predictable path based on pid and time
    let pid = std::process::id();
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
    let path = format!("/tmp/rust_demo_{}_{}.tmp", pid, now);
    // create file without setting restrictive permissions
    let _ = File::create(&path); // ignore errors intentionally (bad practice)
    path
}

// ----------------------- 4) Command invocation via shell (command injection risk) ---
fn list_files_shell(user_arg: &str) {
    // Unsafe pattern: constructing a shell command with user-controlled input
    // Using "sh -c" is commonly flagged by SAST rules.
    let cmd = format!("ls -l {}", user_arg); // potential injection if user_arg contains metacharacters
    // spawn with sh -c on purpose to show pattern
    let _ = Command::new("sh")
        .arg("-c")
        .arg(cmd)
        .status(); // intentionally ignore Result (bad practice)
}

// ----------------------- 5) Unchecked indexing via get_unchecked (UB risk) -----------
fn unchecked_index_example(vec: &[u8], idx: usize) -> u8 {
    unsafe {
        // using get_unchecked bypasses bounds checks — subtle UB if idx invalid
        *vec.get_unchecked(idx)
    }
}

// ----------------------- 6) Unsafe pointer cast + reinterpretation -------------------
fn unsafe_pointer_reinterpret() {
    // create a small integer and write beyond its size via raw pointer arithmetic
    let mut x: u32 = 0x41414141;
    let p: *mut u8 = &mut x as *mut u32 as *mut u8;
    unsafe {
        // write bytes directly — depending on endianness this corrupts memory
        // subtle UB if misused; static analyzers should flag raw pointer ops
        std::ptr::write(p, 0xFFu8);
    }
    // print to appear normal
    println!("x after raw write: 0x{:08x}", x);
}

// ----------------------- 7) Double free via Box::from_raw (UB) -----------------------
fn double_free_demo() {
    // Create a box, convert to raw, then create two Boxes from the same raw pointer and drop both
    let boxed = Box::new(123u32);
    let raw = Box::into_raw(boxed);
    unsafe {
        // first reconstitute and drop
        let b1 = Box::from_raw(raw);
        drop(b1);
        // second reconstitute -> double free (UB)
        let b2 = Box::from_raw(raw);
        drop(b2);
    }
    // Program might continue but this is undefined behavior and should be flagged.
}

// ----------------------- 8) Returning reference to leaked memory (looks safe) -------
fn leak_and_return() -> &'static str {
    // Leak a String to obtain a 'static str — not UB but hides memory management issues
    let s = String::from("leaked_string");
    let leaked: &'static str = Box::leak(s.into_boxed_str());
    leaked
}

// ----------------------- 9) Swallowing errors (bad error handling) ------------------
fn swallow_errors_demo(path: &str) {
    // intentionally ignoring Result and errors
    let mut f = match File::open(path) {
        Ok(f) => f,
        Err(_) => return, // swallow
    };
    let mut buf = String::new();
    let _ = f.read_to_string(&mut buf); // ignore read errors
    // intentionally not handling content
    let _ = buf;
}

// ----------------------- 10) Excessive use of unwrap/expect (clippy: unwrap_used) ----
fn unwrap_demo() {
    // multiple unwraps look normal but are brittle; clippy flags unwrap_used
    let args: Vec<String> = env::args().collect();
    // if not provided, unwrap will panic in production
    let maybe = args.get(1).unwrap_or(&String::from("default")).clone();
    println!("arg: {}", maybe);
    // parse number unsafely
    let n: i32 = "42".parse().expect("should parse"); // expect_used
    println!("parsed: {}", n);
}

// ----------------------- 11) needless_clone + redundant allocation (clippy) ----------
fn needless_clone_demo(s: &str) {
    // Here clone is unnecessary; clippy warns about needless_clone
    let x = s.to_string();
    let y = x.clone(); // needless clone
    println!("cloned len: {}", y.len());
}

// ----------------------- 12) Weak "obfuscation" used as pseudo-crypto ---------------
fn weak_xor_obfuscate(s: &str) -> String {
    s.bytes().map(|b| (b ^ 0xAAu8) as char).collect()
}

// ----------------------- main: tie things together, look benign ----------------------
fn main() {
    // Intentionally non-rustfmt formatting to test rustfmt detection:
    println!("Subtle Rust vulnerabilities demo - looks normal");

    // 1: secret printed (should be caught by secret scanners)
    println!("api secret preview: {}...", &API_SECRET[..6]);

    // 2: weak token
    println!("weak token: {}", weak_token());

    // 3: predictable tmp file
    let tmp = predictable_tmpfile();
    println!("predictable tmp created: {}", tmp);

    // 4: list files using shell with safe-looking arg
    let safe_arg = ".";
    list_files_shell(safe_arg);

    // 5: unchecked index but using small index so usually fine
    let v = vec![10u8, 20, 30];
    let val = unsafe { unchecked_index_example(&v, 1) };
    println!("unchecked index result: {}", val);

    // 6: unsafe pointer reinterpret
    unsafe_pointer_reinterpret();

    // 7: double free demo — may corrupt heap silently
    // Run in a confined test; often program continues but UB is present
    double_free_demo();

    // 8: leaked static string (hides memory management)
    let leaked = leak_and_return();
    println!("leaked static: {}", leaked);

    // 9: swallow errors on a likely-missing path
    swallow_errors_demo("/does/not/exist.txt");

    // 10: unwrap demo (panic risk)
    unwrap_demo();

    // 11: needless clone
    needless_clone_demo("hello");

    // 12: weak obfuscation
    println!("xor obf: {}", weak_xor_obfuscate("secret"));

    // Exit normally
    println!("Done (program terminated normally)");
}
